craft_parts/executor/part_handler.py

```diff
@@ -187,6 +194,26 @@ def _migrate(
             src_path = str(Path(sub_path) / f)
             self.migrated_directories[dst_partition][src_path] = f
 
+    def _filter_already_distributed(
+        self, visible_files: set[str], visible_dirs: set[str]
+    ) -> tuple[set[str], set[str]]:
+        """Exclude files in sub path already distributed to another partition."""
+        files: set[str] = set()
+        dirs: set[str] = set()
+        if not self._sub_paths_distributed:
+            return visible_files, visible_dirs
```

Comment:
We can also move this block to the top of this method, before declaring `files` and `dirs`.